Scalar documentation (#2357)
graphql-dotnet/graphql-dotnet: docs2/site/docs/getting-started/custom-scalars.md

- Literal Parsing: Transforms a scalar from its client-side representation as an argument to its server-side representation.

Parsing for arguments and variables are handled separately because while arguments must always be expressed in GraphQL

> query syntax, variable format is transport-specific (usually JSON).

Reviewer: ```suggestion
query syntax, variable format is transport-specific (usually JSON). More information about these methods
you can find [here](https://github.com/graphql-dotnet/graphql-dotnet/blob/master/src/GraphQL/Types/Scalars/ScalarGraphType.cs).
```

Author: ```suggestion
query syntax, variable format is transport-specific (usually JSON). You can find more information about
these methods [here](https://github.com/graphql-dotnet/graphql-dotnet/blob/master/src/GraphQL/Types/Scalars/ScalarGraphType.cs).
```

}
```

Vectors should be received in a more structured format:
Vectors should be returned in the same format:

```json
{
    "data": {

>         "getVector": {

Reviewer: Strange. Is it a scalar? This is an object. Documentation was incorrect?

Author: No, the documentation was demonstrating how to have a scalar return a complex object.  I moved the demonstration further down the page, after a more basic demonstration.

}
```

Vectors should be returned in the same format:

```json
{
    "data": {
        "getVector": "23, 43, 66"
    }
}
```

> ### 1. Create the class for the server-side representation.

Reviewer: ```suggestion
### 1. Create the class/struct for the server-side representation.
```

- `IntValue` - integers that can be represented within an `int`
- `LongValue` - integers that can be represented within a `long`
- `BigIntValue` - integers that can be represented within a `BigInteger`
- `FloatValue` - floating-point values that can be represented within a `double`
- `DecimalValue` - floating-point values that can be represented within a `decimal`
- `StringValue` - string values
- `EnumValue` - enumeration values
- `NullValue` - representing `null` - must be handled by all scalars

If your custom scalar accepts floating-point values, you must be sure to handle all 5 of the
numeric types, since queries like `{ field(arg: 3) }` is parsed as an `IntValue` even though
it is also a valid floating-point number.

In the sample below, only `NullValue` and `StringValue` need to be handled.

For any type that is not handled, or when the value cannot be parsed, you must throw an exception.
`ThrowLiteralConversionError` is provided as a convenient method to facilitate throwing an exception
when the type does not match.

```csharp
//in Vector3Type

public override object ParseLiteral(IValue value)
{
    if (value is NullValue)
        return null;

>     if (value is StringValue stringValue) return ParseValue(stringValue.Value);

Reviewer: ```suggestion
    if (value is StringValue stringValue)
        return ParseValue(stringValue.Value);
```

is handled within `ParseValue`. In this case, serialization of `Vector3` is serialized as
a string, which matches the expected data type within `ParseValue`.

```csharp
// In Vector3Type

public override object Serialize(object value)
{
    if (value == null)
        return null;

    if (value is Vector3 vector3)
    {
        return $"{vector3.X},{vector3.Y},{vector3.Z}";
    }

    return ThrowSerializationError(value);
}
```

### 7. Override `ScalarGraphType.ToAST` if necessary.

The infrastructure converts default field values to AST representations during initialization
in order to verify that the default values are valid within an AST tree. The default implementation
calls `Serialize` to convert the value to its client-side equivalent, then embeds it into an
appropriate AST node based on the data type returned from `Serialize`. It is likely you will
only need to override this method if you are creating a custom scalar that returns enumeration
values, or if you are returning structured data.

### 8. Code-first: Register the .NET type mapping with the schema (optional).

In order to allow simple field mappings like `Field(x => x.Vector)`, you need to register the

> custom scalar with the schema so that the proper graph type can be selected for the field.

Reviewer: ```suggestion
custom scalar within the schema so that the proper graph type can be selected for the field.
```
?

> Call `RegisterTypeMapping` with the CLR data model type and the graph type you want it mapped to.

Reviewer: ```suggestion
Call `RegisterTypeMapping` with the CLR data model type and the graph type you want it map to.
```

?

Author: No. "..it mapped to" is correct.  "it to map to" would also be correct, but "it map to" is not.

Reviewer: ok

```csharp
public class MySchema
{
    public MySchema()
    {
        ...

        this.RegisterTypeMapping<Vector3, Vector3Type>();
    }
}
```

This is not necessary if you use the alternate `Field<T>` syntax which specifies the graph type

> to be used for the field.

Reviewer: ```suggestion
to be used for the field or if your scalar data type is marked with `GraphQLMetadata` attribute
setting `InputType` and/or `OutputType` properies.
```

In this example, you created a custom scalar. In summary:

> - Create a data class for the server-side representation of the scalar

Reviewer: ```suggestion
- Create (if necessary) a data class/struct for the server-side representation of the scalar
```

- Implement a `ScalarGraphType` which handles variable parsing, literal parsing, and serialization
- Register the `ScalarGraphType` within the DI container

You can also choose to override `CanParseLiteral`, `CanParseValue` or `IsValidDefault` for
enhanced performance. The default implementations call `ParseLiteral`, `ParseValue` and
`ToAST` respectively, returning `false` if an exception is caught, or `true` otherwise.
If you do choose to implement these methods, note that those methods must not throw an
exception, and that they are not always called when executing a document.

## Vector3 sample with combined string/structured parsing and serialization

Keep in mind that the serialized value returned by custom scalar can by anything that the
environment allows. For example it can be a structured object, rather than a simple value.

> This is also true of variables, however literals can only be the represented by the native

Reviewer: ```suggestion
This is also true for variables, however literals can only be the represented by the native
```

scalar types supported by the GraphQL query language -- integers, floating-point values,

> booleans and null values. You cannot parse a structured literal object into a scalar value.

Reviewer: > You cannot parse a structured literal object into a scalar value.

That's a question. I think it's possible and in general does not violate logic.

So to extend our sample, let's assume that we want the Vector3 scalar to instead accept and
return data in a more structured format, in addition to supporting the string format for literals.

Here is a sample of a variable supporting a more structured format:

```graphql
mutation AddVector($vector3: Vector3!) {
    addVector(vector3: $vector3)
}

//variables
{
    "vector3": {
        "X":"23",
        "Y":"43",
        "Z":"66"
    }
}
```

And a sample of a response with a vector in a more structured format:

```json
{
    "data": {
        "getVector": {

>             "X":"23",

Reviewer: By the way... should we change "X" to "x" in all requests/responses?

Author: Yes

            "Y":"43",
            "Z":"66"
        }
    }
}
```

> ### 1. Change `ParseValue` to accept strings or structured data:

Reviewer: The same for `ParseLiteral` 🤔

}
```

Once the raw string is extracted from the value node, normal parsing can proceed.

### 5. Prepare to accept `Vector3` inputs from query variables. Implement `ScalarGraphType.ParseValue`.

Similar to `ParseLiteral`, you must keep in mind the expected format of values that are likely
to be presented to this method. For instance, if you are using a JSON deserializer, you may be
presented with values of any of these types (or more, depending on your deserializer configuration):

- `int`
- `long`
- `ulong`
- `BigInteger`
- `double`
- `decimal`
- `string`
- `null` - must be handled by all scalars

On top of that, if you are calling this method from `ParseLiteral`, you must handle the types
passed from it. So if your scalar needs to handle floating-point values, you likely need to handle
`int`, `long`, `ulong`, `BigInteger`, `double` and `decimal` types.

For any type that is not handled, or when the value cannot be parsed, you must throw an exception.
`ThrowValueConversionError` is provided as a convenient method to facilitate throwing an exception
when the type does not match.

In the sample below, only `null` and `string` types need to be handled.

```csharp

> // In Vector3Type
> public override object ParseValue(object value)

Reviewer: ```suggestion
// In Vector3Type

public override object ParseValue(object value)
```

{
    public MySchema()
    {
        ...

        this.RegisterTypeMapping<Vector3, Vector3Type>();
    }
}
```

This is not necessary if you use the alternate `Field<T>` syntax which specifies the graph type
to be used for the field, or if your scalar data type is marked with the `GraphQLMetadata`
attribute setting the `InputType` and/or `OutputType` properties.

In this example, you created a custom scalar. In summary:

- Create (if necessary) a data class/struct for the server-side representation of the scalar
- Implement a `ScalarGraphType` which handles variable parsing, literal parsing, and serialization
- Register the `ScalarGraphType` within the DI container

You can also choose to override `CanParseLiteral`, `CanParseValue` or `IsValidDefault` for
enhanced performance. The default implementations call `ParseLiteral`, `ParseValue` and
`ToAST` respectively, returning `false` if an exception is caught, or `true` otherwise.
If you do choose to implement these methods, note that those methods must not throw an
exception, and that they are not always called when executing a document.

## Vector3 sample with combined string/structured parsing and serialization

> Keep in mind that the serialized value returned by custom scalar can by anything that the

Reviewer: ```suggestion
Keep in mind that the serialized value returned by custom scalar can be anything that the
```

            throw new FormatException($"Failed to parse {nameof(Vector3)} from input '{vector3InputString}'. Input should be a string of three comma-separated floats in X Y Z order, ex. 1.0,2.0,3.0");
        }
    }

    return ThrowValueConversionError(value);
}
```

> 5. Implement `ScalarGraphType.Serialize` so `Vector3` instances can be sent to the client.

Reviewer: ```suggestion
6. Implement `ScalarGraphType.Serialize` so `Vector3` instances can be sent to the client.
```

> Keep in mind that a null value also required to be handled.

Reviewer: ```suggestion
Keep in mind that a `null` value also required to be handled.
```

I suggest to highlight `null` everywhere in the text.

It is highly recommended that the type of data this method returns match the same type as
is handled within `ParseValue`. In this case, serialization of `Vector3` is serialized as
a string, which matches the expected data type within `ParseValue`.

```csharp
// In Vector3Type

public override object Serialize(object value)
{
    if (value == null)
        return null;

    if (value is Vector3 vector3)
    {
        return $"{vector3.X},{vector3.Y},{vector3.Z}";
    }

    return ThrowSerializationError(value);
}
```

> 6. Override `ScalarGraphType.ToAST` if necessary.

Reviewer: ```suggestion
7. Override `ScalarGraphType.ToAST` if necessary.
```

- Register the `ScalarGraphType` with the DI container

You can also choose to override `CanParseLiteral`, `CanParseValue` or `IsValidDefault` for
enhanced performance. The default implementations call `ParseLiteral`, `ParseValue` and
`ToAST` respectively, returning `false` if an exception is caught, or `true` otherwise.
If you do choose to implement these methods, note that those methods must not throw an
exception, and that they are not always called when executing a document.

## Vector3 sample with combined string/structured parsing and serialization

Now to extend our sample, let's assume that we want the Vector3 scalar to instead accept and
return data in a more structured format, in addition to support the string format for variables.

Here is a sample of a variable supporting a more structured format:

```graphql
mutation AddVector($vector3: Vector3!) {
    addVector(vector3: $vector3)
}

//variables
{
    "vector3": {
        "X":"23",
        "Y":"43",
        "Z":"66"
    }
}
```

And a sample of a response with a vector in a more structured format:

```json
{
    "data": {

>         "getVector": {

Reviewer: I'm not sure that this is the right example (although he was here before). Scalar is to determine the leaf nodes of the request/response tree. Yes, custom scalar can have some kind of formatting, but they cannot add additional levels in the tree. These are no longer scalar, but objects.

Author: The old 'custom scalars' page explicitly gave this sample, with a structured object being returned.

https://graphql-dotnet.github.io/docs/getting-started/custom-scalars

🤷‍♂️

I actually think it's a pretty neat sample, and demonstrates the power of custom scalars, including a need for `ToAST`.  But maybe we should add a note that the resulting schema is not spec-compliant.

Or perhaps we come up with another sample that can demonstrate a need for `ToAST`, and remove this one.

> These are no longer scalar, but object

True...  but whose fault is that?

Do we say that serialization and variables are outside the scope of graphql (as per spec), and therefor Vector3 is a scalar?  Along that logic, we look and see that ParseValue is explicitly looking for an object (dictionary) rather than a Vector3.  Why?  Because the json serialization engine cannot properly handle deserialization of Vector3.  Nor serialization for that matter (since it gets serialized as an object).  So we added a patch to ParseValue to patch a broken deserialization engine (e.g. `System.Text.Json`) that does not support deserializing Vector3 natively.

Or do we say that the graphql scalar should be written from the premise that a json serialization engine is used?  Then we can say that we should not be returning a Vector3 from Serialize because that type will be serialized as an object, which is "illegal" from graphql's POV.  (It's not a scalar anymore.)

We can even say that we "invented" a custom transport -- MyJSON -- where `{x:1,y:2,z:3}` is a Vector3 scalar.

I would generally write all of the code from the POV that a json serialization engine will be used.  Which is why I would remove native types (e.g. Guid or Uri) from being handled by ParseValue.  But in those cases at least it is a valid external representation of the value, and no harm will come from supporting it.

Anyway, back to the original question: I think we should provide a sample of a custom scalar with a reasonable premise and enough complexity so that the tooling can be demonstrated well.  It does not have to be a complex version of Vector3 of course.  But hopefully something that requires a `ToAST` implementation.

Reviewer: > But maybe we should add a note that the resulting schema is not spec-compliant.

Probably. See http://spec.graphql.org/June2018/#sec-Scalars:
![изображение](https://user-images.githubusercontent.com/21261007/110234414-689f4b00-7f3b-11eb-9527-a6bb66abe484.png)

> True... but whose fault is that?

I would say ours, because we show people how to use scalar incorrectly.

> Or do we say that the graphql scalar should be written from the premise that a json serialization engine is used?

GraphQL response is **always** JSON.

> Which is why I would remove native types (e.g. Guid or Uri) from being handled by ParseValue. But in those cases at least it is a valid external representation of the value, and no harm will come from supporting it.

We should not delete them.

> But hopefully something that requires a ToAST implementation.

I already thought about JSON scalar. It will take some kind of library to implement it, so it cannot be placed in the main project. We can place that scalar in our projects for serialization. But for now just try to make an example.

Reviewer: https://github.com/graphql/graphql-spec/issues/826

Author: > GraphQL response is always JSON.

And variables is not?

Author: Well, we can simply delete the second half of the sample.

Author: > GraphQL response is always JSON.

Not according to the specification.

http://spec.graphql.org/June2018/#sec-JSON-Serialization

> JSON is the most common serialization format for GraphQL. Though as mentioned above, GraphQL does not require a specific serialization format.

Author: The GraphQL spec defines the GraphQL query language, but does not make any requirements on the format of variables or result.  JSON is the most common format for variables and results to be formatted as.  I suggest limiting us to only JSON.  Then Guid and Uri and Enum are types that are not supported by JSON and hence have no need to be included within ParseValue.  And returning an object is invalid for a scalar to do.

But if we are designing the library where JSON is not a required input/output format -- for instance, perhaps the binary serializer within `System.Runtime.Serialization` was used -- then perhaps Guid and Uri and Enum are types that should be supported by ParseValue.  But also Vector3 is a completely valid scalar value, just as much as Guid or Uri.  Just like Uri, it can be serialized to a string, or can be split into parts (e.g. Scheme, Host, Port, Path, Query, etc).  There is no requirement that it MUST be serialized to a string.  It is simply a scalar that can be serialized by the binary transport in its native format.

So we can design this from a POV that only JSON transports are used, or not.

Reviewer: Good news in https://github.com/graphql/graphql-spec/issues/826 It seems that it is quite possible to serialize the scalar to objects:

> the value returned by custom scalar for the result coercion (in GraphQL.js this function is called serialize) can by anything that the Programming language/environment allows. For example it can be a Money object.

> GraphQL in general doesn't not limit at all this aspect: it is an implementation detail of the GraphQL implementation.

> To phrase it a bit different: The result of an GraphQL execution is tree, where Scalars or Enums are the leaves from a type perspective. But what the actual values of these types are at runtime can't be predicted or described by the spec.

I suggest you copy these phrases as is into documentation (with some changes). You can do this at the very beginning of a more complex second example to explain the motivation.

Please also add somewhere in text links to https://www.graphql-scalars.com/scalars-guide and https://www.graphql.de/blog/scalars-in-depth/ .

6. Implement `ScalarGraphType.Serialize` so `Vector3` instances can be sent to the client.
Keep in mind that a `null` value also required to be handled.

It is recommended that the type of data this method returns match the same type as
is handled within `ParseValue`. In this case, serialization of `Vector3` is serialized as
a string, which matches the expected data type within `ParseValue`.

```csharp
// In Vector3Type

public override object Serialize(object value)
{
    if (value == null)
        return null;

    if (value is Vector3 vector3)
    {
        return $"{vector3.X},{vector3.Y},{vector3.Z}";
    }

    return ThrowSerializationError(value);
}
```

7. Override `ScalarGraphType.ToAST` if necessary.

The infrastructure converts default field values to AST representations during initialization
in order to verify that the default values are valid within an AST tree. The default implementation
calls `Serialize` to convert the value to its client-side equivalent, then embeds it into an
appropriate AST node based on the data type returned from `Serialize`. It is likely you will
only need to override this method if you are creating a custom scalar that returns enumeration

> values, or if you are returning structured data.

Reviewer: > returning _structured_ data

Somewhere here may be note about the nature of scalars - scalars are not required to appear as primitive types i.e. GraphQL "supports" composite scalars on _serialization level_. See https://github.com/graphql-dotnet/graphql-dotnet/pull/2357#discussion_r589441410

Reviewer: Done?

Author: I guess I don't know what I should do here.  We have documentation and an extensive sample regarding structured data a few paragraphs below.

Reviewer: ok

In this example, you created a custom scalar. In summary:

> - Create a class for the server-side representation of the scalar

Reviewer: What class ? ScalarGraphType or data-class ?

Author: Yeah I didn't write this...

Author: ```suggestion
- Create a data class for the server-side representation of the scalar
```

- Implement a `ScalarGraphType` which handles variable parsing, literal parsing, and serialization

> - Register the `ScalarGraphType` with the DI container

Reviewer: ```suggestion
- Register the `ScalarGraphType` within the DI container
```
?

You can also choose to override `CanParseLiteral`, `CanParseValue` or `IsValidDefault` for
enhanced performance. The default implementations call `ParseLiteral`, `ParseValue` and
`ToAST` respectively, returning `false` if an exception is caught, or `true` otherwise.
If you do choose to implement these methods, note that those methods must not throw an
exception, and that they are not always called when executing a document.

## Vector3 sample with combined string/structured parsing and serialization

Now to extend our sample, let's assume that we want the Vector3 scalar to instead accept and

> return data in a more structured format, in addition to support the string format for variables.

Reviewer: > Somewhere here may be note about the nature of scalars

Or here.

            var x = Convert.ToSingle(dictionary["x"]);
            var y = Convert.ToSingle(dictionary["y"]);
            var z = Convert.ToSingle(dictionary["z"]);
            if (dictionary.Count > 3)
                return ThrowValueConversionError(value);
            return new Vector3(x, y, z);
        }
        catch
        {
            throw new FormatException($"Failed to parse {nameof(Vector3)} from object. Input should be an object of three floats named X Y and Z");
        }
    }

    return ThrowValueConversionError(value);
}
```

2. Change `Serialize` to return structured data.

```csharp
// In Vector3Type

public override object Serialize(object value)
{
    if (value == null)
        return null;

    if (value is Vector3 vector3)
        return vector3;

    return ThrowSerializationError(value);
}
```

3. Change `ToAST` to return an AST literal that represents the data.

Since `Serialize` no longer returns a type that can be converted to an AST node, it is
necessary to override this method.

```csharp
// In Vector3Type

public override IValue ToAST(object value)
{
    if (value == null)
        return new NullValue();

    if (value is Vector3 vector3)
    {

>         return new StringValue($"{vector3.X},{vector3.Y},{vector3.Z}");

Reviewer: AST should be a string or a structure (object)?

Author: Must be a string, as it needs to be able to be parsed by ParseLiteral.  From what I understand of the graphql spec, and based on the comments in the spec repo, scalar representations within graphql language cannot be structured.  Only the representations as variables or as serialized may be structured.

In my mind, `ParseLiteral` and `ToAST` are inverse functions of each other, and `ParseValue` and `Serialize` are inverse functions of each other.

Reviewer: I think technically there is nothing preventing us to parse `ObjectValue` as literal. What is the difference at least for us parsing input object or scalar from this structured data? Both will work. Right?

Reviewer: I think technically there is nothing preventing us to parse `ObjectValue` as literal. What is the difference at least for us parsing input object or scalar from this structured data? Both will work. Right?

Reviewer: Try one more test supplying custom scalar as structured (object) literal.

Author: Have not tried it yet, and I think it would work, except pretty sure it fails validation (as it should). It would be interesting to try though.

Reviewer: Yes, interesting.

Author: It works!!!  Tests added.

Reviewer: 👍

    }

    return ThrowASTConversionError(value);
}
```

With these changes, a variable can be parsed as a string or as a structured object, and is always returned as a structured object.

## ValueConverter

When `GetArgument<T>` is called, the argument value is coerced to the requested type via the `ValueConverter`.
No conversion takes place when the requested type matches the type of the object or scalar (the type returned from
`ParseLiteral` or `ParseValue`). But you can also use the value converter to assist with input deserialization.

For instance, you may be using `IdGraphType` within your schema as unique identifiers for your data objects. Pursuant
to the GraphQL specification, these identifiers should be passed as strings such as in the below example:

```graphql
{
    widget (id: "3") {
        name
    }
}
```

However, your code may use integer identifiers in the data access layer. So when you call `context.GetArgument<int>("id")`,
GraphQL.NET calls the value converter to convert the string to an integer.

> The value converter can be extended globally by calling the static method `Register` as follows:

Reviewer: This API is one to think about changing from static to instance method (v5).

Author: Or delete it altogether...

Reviewer: Resolvefieldcontext.GetArgument still should be rather agile.

```csharp
ValueConverter.Register<Vector3, string>(v => $"{v.X},{v.Y},{v.Z}");
```

The above method registers a conversion from the `Vector3` format to a `string`. Since the registration is static,
it should only be done once per application lifetime. For instance, in a static constructor of your schema.

```csharp
public class MySchema : Schema
{
    static MySchema()
    {
        ValueConverter.Register<Vector3, string>(v => $"{v.X},{v.Y},{v.Z}");
    }

    ...
}
```

## Null values

> Custom scalars process and handle null values during serialization and deserialization. This allows

Reviewer: > This allows
for custom scalars that can assist when you have database values such as 0 that should represent null
when exposed outside the schema.

It is difficult for me to understand. There is no mistake here?

Author: I have a database where that is the case.  Another scenario I run into often is null vs empty strings.  Specifically, some databases hold null where they should represent empty strings, or some databases that hold empty strings where they should represent null.

If you have a suggestion of how I can explain this better, or a better sample to show altogether, I'm listening.

Reviewer: I'm talking about a long sentence to read and understand to me as not a native speaker. Maybe split to shorter ones?

Reviewer: Two that, two when.

Author: See suggestion here: https://github.com/graphql-dotnet/graphql-dotnet/pull/2357/files#r593837457

Modify or merge as desired.

Reviewer: done

for custom scalars that can assist when you have database values such as 0 that should represent null
when exposed outside the schema. Below is an example of a scalar intended to represent a database
autoincrementing numeric identifier internally, where null values are stored as 0.

```csharp
public class DbIdGraphType : ScalarGraphType
{
    public DbIdGraphType()
    {
        Name = "DbId";
    }

    public override object ParseLiteral(IValue value) => value switch
    {
        StringValue s => int.TryParse(s.Value, out int i) && i > 0 ? i : throw new FormatException($"'{s.Value}' is not a valid identifier."),

>         null => 0,

Reviewer: `null` or `NullValue` or both ?

Author: ```suggestion
        NullValue _ => 0,
```
You're right; `NullValue` is proper.  `value` cannot be `null` here.

        _ => ThrowLiteralConversionError(value)
    };

    public override object ParseValue(object value) => value switch
    {
        string s => int.TryParse(s, out int i) && i > 0 ? i : throw new FormatException($"'{s}' is not a valid identifier."),
        null => 0,
        _ => ThrowValueConversionError(value)
    };

    public override object Serialize(object value) => value switch
    {
        int i => i > 0 ? i.ToString() : i == 0 ? null : ThrowSerializationError(value),
        _ => ThrowSerializationError(value)
    };
}
```

## Replacing built-in scalar types

In some cases you may want or need to replace the functionality of the built-in graph types. This can
be accomplished by running a schema node visitor on the schema after it has been initialized. Keep in
mind that the node visitor will walk through the introspection portion of the schema, so replacing
a built-in type may affect the operation of introspection queries.

> Below is a sample of how to replace the built-in `IdGraphType` with the `DbIdGraphType` shown above.

Reviewer: The thought is correct, but I'm not sure yet it will work.

Author: I added a test to verify that it works.

Author: Sorry, I added a test to show that the `DbIdGraphType` works with nullability coercion.

Author: This sample was replaced with a more common scenario -- replacing the boolean graph type (which is more strict in 4.0) to a version that allows coercion from any type of value (similar to how it was supported in 3.x).

Reviewer: Ok

1. Name the `DbIdGraphType` to be the same as the built-in type

```csharp
public class DbIdGraphType : ScalarGraphType
{
    public DbIdGraphType()
    {
        Name = "ID";
    }

    ...
}
```

2. Create a node walker

```csharp

> // sungam3r to provide sample

Reviewer: https://github.com/graphql-dotnet/graphql-dotnet/issues/1952#issuecomment-786177240  Found!

// perhaps we should provide a node walker for this purpose, or provide a more convenient method to replace types

> // note: don't we need to fix up schematypes also so that if a type name is provided to a query variable argument, it gets a proper reference?

Reviewer: This is an interesting question. Perhaps not because _replacing_  involves replacement of behavior but not name. Name should stay the same.

Author: We are testing for this scenario with the solution in #2388 as it tests variables, literals, and serialization, for both schema-first and code-first.

Reviewer: So the names match. Right?

Author: Right.

Author: Sample replaced, so resolving this not-applicable comment.

// todo: add validation for sample
```

3. Execute the node walker on the schema after it builds, during the validation stage

```csharp
public class MySchema : Schema
{
    ...

    private readonly MyNodeWalker _myNodeWalker = new MyNodeWalker();
    protected override void Validate()
    {

>         _myNodeWalker.Run(this);

Reviewer: Before or after `base.Validate`? 🤔

Author: Removed.

            var y = Convert.ToSingle(dictionary["y"]);
            var z = Convert.ToSingle(dictionary["z"]);
            if (dictionary.Count > 3)
                return ThrowValueConversionError(value);
            return new Vector3(x, y, z);
        }
        catch
        {
            throw new FormatException($"Failed to parse {nameof(Vector3)} from object. Input should be an object of three floats named X Y and Z");
        }
    }

    return ThrowValueConversionError(value);
}
```

### 2. Change `Serialize` to return structured data.

```csharp
// In Vector3Type

public override object Serialize(object value)
{
    if (value == null)
        return null;

    if (value is Vector3 vector3)
        return vector3;

    return ThrowSerializationError(value);
}
```

### 3. Change `ToAST` to return an AST literal that represents the data.

Since `Serialize` no longer returns a type that can be converted to an AST node, it is
necessary to override this method.

```csharp
// In Vector3Type

public override IValue ToAST(object value)
{
    if (value == null)
        return new NullValue();

    if (value is Vector3 vector3)
    {

>         return new StringValue($"{vector3.X},{vector3.Y},{vector3.Z}");

Reviewer: ```suggestion
        return new ObjectValue(
            new ObjectField("x", new FloatValue(vector3.X)),
            new ObjectField("y", new FloatValue(vector3.Y)),
            new ObjectField("z", new FloatValue(vector3.Z)),
        );
```

?

Reviewer: And if so please add new constuctor in `ObjectValue` with `params` keyword.

    }

    return ThrowASTConversionError(value);
}
```

With these changes, a variable can be parsed as a string or as a structured object, and is always returned

> as a structured object. Again, note that as an input literal, it must be parsed as a string.

Reviewer: > Again, note that as an input literal, it must be parsed as a string.

I doubt about that (_must_).

## ValueConverter

When `GetArgument<T>` is called, the argument value is coerced to the requested type via the `ValueConverter`.
No conversion takes place when the requested type matches the type of the object or scalar (the type returned from
`ParseLiteral` or `ParseValue`). But you can also use the value converter to assist with input deserialization.

For instance, you may be using `IdGraphType` within your schema as unique identifiers for your data objects. Pursuant

> to the GraphQL specification, these identifiers should be passed as strings such as in the below example:

Reviewer: Should they?
![изображение](https://user-images.githubusercontent.com/21261007/111027643-89b3e000-8402-11eb-8c7a-ba659fbe9bc4.png)

Author: ```suggestion
to the GraphQL specification, these identifiers may be passed as strings such as in the below example:
```

```graphql
{
    widget (id: "3") {
        name
    }
}
```

However, your code may use integer identifiers in the data access layer. So when you call `context.GetArgument<int>("id")`,
GraphQL.NET calls the value converter to convert the string to an integer.

The value converter can be extended globally by calling the static method `Register` as follows:

```csharp
ValueConverter.Register<Vector3, string>(v => $"{v.X},{v.Y},{v.Z}");
```

> The above method registers a conversion from the `Vector3` format to a `string`. Since the registration is static,

Reviewer: ```suggestion
The above method registers a conversion from the `Vector3` struct to a `string`. Since the registration is static,
```

it should only be done once per application lifetime. For instance, in a static constructor of your schema.

```csharp
public class MySchema : Schema
{
    static MySchema()
    {
        ValueConverter.Register<Vector3, string>(v => $"{v.X},{v.Y},{v.Z}");
    }

    ...
}
```

## Null values

Custom scalars process and handle null values during serialization and deserialization. This allows
for custom scalars that can assist when you have database values such as 0 that should represent null
when exposed outside the schema. Below is an example of a scalar intended to represent a database
autoincrementing numeric identifier internally, where null values are stored as 0.

```csharp
public class DbIdGraphType : ScalarGraphType
{
    public DbIdGraphType()
    {
        Name = "DbId";
    }

    public override object ParseLiteral(IValue value) => value switch
    {

>         StringValue s => int.TryParse(s.Value, out int i) && i > 0 ? i : throw new FormatException($"'{s.Value}' is not a valid identifier."),

Reviewer: So passing 0 in inputs is forbidden?

Author: That's what is demonstrated, yes.

    };

    public override object Serialize(object value) => value switch
    {
        int i => i > 0 ? i.ToString() : i == 0 ? null : ThrowSerializationError(value),
        _ => ThrowSerializationError(value)
    };
}
```

## Replacing built-in scalar types

In some cases you may want or need to replace the functionality of the built-in graph types. This can
be accomplished by registering a replacement scalar before the schema has been initialized. Keep in
mind that replacing a built-in type may affect the operation of introspection queries.

In order to replace a built-in scalar graph type, the new scalar graph type must:

1. inherit from the scalar graph type it is replacing; and
2. have the `Name` property set to the name of the built-in graph type.

You may then override any of the members to provide custom implementations. Note that most of the
built-in scalars override `CanParseLiteral`, so it may be necessary to override that method if you
override `ParseLiteral`. Check the source code for the built-in scalar type you are overriding for
further reference.

Below is a sample of how to replace the built-in `BooleanGraphType` so it will accept 0 and non-zero

> values to represent false and true.

Reviewer: ```suggestion
values to represent `false` and `true`.
```

### 1. Create a new scalar graph type `MyBooleanGraphType`. Inherit from `BooleanGraphType` and set
the name to be `Boolean`.

```csharp
public class MyBooleanGraphType : BooleanGraphType
{
    public MyBooleanGraphType()
    {
        Name = "Boolean";
    }
}
```

> ### 2. Override the methods as necessary; in this case we must override all of them except `IsValidDefault`.

Reviewer: ```suggestion
### 2. Override the methods as necessary; in this case we must override all of them except `Serialize` and `IsValidDefault`.
```

```csharp
public class MyBooleanGraphType : BooleanGraphType
{
    public MyBooleanGraphType()
    {
        Name = "Boolean";
    }

    public override object ParseLiteral(IValue value) => value switch
    {
        BooleanValue b => b.Value,
        IntValue i => ParseValue(i.Value),
        LongValue l => ParseValue(l.Value),
        BigIntValue bi => ParseValue(bi.Value),
        StringValue s => ParseValue(s.Value),
        FloatValue f => ParseValue(f.Value),
        DecimalValue d => ParseValue(d.Value),
        NullValue _ => null,
        _ => ThrowLiteralConversionError(value)
    }

>     public virtual bool CanParseLiteral(IValue value)

Reviewer: ```suggestion
    public override bool CanParseLiteral(IValue value)
```

    {
        try
        {
            _ = ParseLiteral(value);
            return true;
        }
        catch
        {
            return false;
        }
    }

    public override object ParseValue(object value) => value switch
    {
        bool _ => value,
        byte b => b != 0,
        sbyte sb => sb != 0,
        short s => s != 0,
        ushort us => us != 0,
        int i => i != 0,
        uint ui => ui != 0,
        long l => l != 0,
        ulong ul => ul != 0,
        BigInteger bi => bi != 0,
        float f => f != 0,
        double d => d != 0,
        decimal d => d != 0,
        string s => bool.Parse(s),
        null => null,
        _ => ThrowValueConversionError(value)
    }

>     public virtual bool CanParseValue(object value)

Reviewer: ```suggestion
    public override bool CanParseValue(object value)
```

    {
        try
        {
            _ = ParseValue(value);
            return true;
        }
        catch
        {
            return false;
        }
    }

>     public virtual object Serialize(object value) => ParseValue(value);
> 

Reviewer: ```suggestion
```

from "base.base"

    public override IValue ToAST(object value) => Serialize(value) switch
    {
        bool b => new BooleanValue(b),
        null => new NullValue(),
        _ => ThrowASTConversionError(value)
    };
}
```

### 3. Register the custom scalar within your schema.

> The final step is to register an instance of the custom scalar to the schema. This can be

Reviewer: ```suggestion
The final step is to register an instance of the custom scalar within the schema. This can be
```

done for code-first or schema-first schemas. For code-first schemas, register it within
your constructor via `RegisterType`, as follows:

```csharp
public class MySchema : Schema
{
    public void MySchema()
    {
        Query = ....;

        RegisterType(new MyBooleanGraphType());
    }
}
```

> For schema-first schemas, register it immediately after calling `Schema.For` to create the schema.

Reviewer: ```suggestion
For schema-first schemas, register it immediately after calling `Schema.For` to create the schema.
Immediately after calling `Schema.For` the schema is not yet initialized, therefore allows registration of types.
```

```csharp
var schema = Schema.For(...);
schema.RegisterType(new MyBooleanGraphType());
```

Now all `BooleanGraphType` references in your schema will utilize the new `MyBooleanGraphType`

> registered with the schema. This technique can be used to replace any of the built-in graph types.

Reviewer: ```suggestion
registered within the schema. This technique can be used to replace any of the built-in graph types.
```

Note that if you set the `ResolvedType` property of a field or argument to an instance of a built-in

> type, or provide an instance of a built-in type to an applicable constructor, it will not be replaced

Reviewer: > it will **not be** replaced with your registered replacement built-in type
> This is **by design**.

😮 It was a surprise for me. Why is that?

Author: Setting the `ResolvedType` is, by design, supposed to use the exact instance of the graph type that was passed to it.  This is necessary when writing code like this:

```csharp
var s = new Schema();
var query = new ObjectGraphType();
query.Name = "Query";
query.Field(...);
s.Query = query;
var mutation = new ObjectGraphType();
mutation.Name = "Mutation";
...
s.Mutation = mutation;
```

Now scalars typically don't have any properties on them, so they don't typically change as in the above example.

But the `RegisterType` is designed to, basically, register a type that can be used when setting all the `ResolvedType` instances throughout the schema.  If an object already has the `ResolvedType` property set, it should not be reset.

So that's why I say that it is by design.  Due to this design, we could change all the introspection types to be 'locked' onto the built-in scalar types by simply setting `ResolvedType` during their initialization, ensuring they always work the same regardless of the user's schema.  The same could be done by users within their own schema if there was a necessary place to do so.

If someone needs to change ALL scalars of a certain type after the schema is built, they can use `ReplaceScalar` afterwards as documented below.

Reviewer: I think I understand. So you are talking about `if ([field/arg].ResolvedType != null) { set_resolved_type }` blocks inside `SchemaTypes.cs`. Right?

Reviewer: Thus, the `ReplaceScalar` method takes its place because it is necessary :)

`ParseLiteral` or `ParseValue`). But you can also use the value converter to assist with input deserialization.

For instance, you may be using `IdGraphType` within your schema as unique identifiers for your data objects. Pursuant
to the GraphQL specification, these identifiers should be passed as strings such as in the below example:

```graphql
{
    widget (id: "3") {
        name
    }
}
```

However, your code may use integer identifiers in the data access layer. So when you call `context.GetArgument<int>("id")`,
GraphQL.NET calls the value converter to convert the string to an integer.

The value converter can be extended globally by calling the static method `Register` as follows:

```csharp
ValueConverter.Register<Vector3, string>(v => $"{v.X},{v.Y},{v.Z}");
```

The above method registers a conversion from the `Vector3` struct to a `string`. Since the registration is static,
it should only be done once per application lifetime. For instance, in a static constructor of your schema.

```csharp
public class MySchema : Schema
{
    static MySchema()
    {
        ValueConverter.Register<Vector3, string>(v => $"{v.X},{v.Y},{v.Z}");
    }

    ...
}
```

## Null values

> Custom scalars process and handle null values during serialization and deserialization. This allows
> for custom scalars that can assist when you have database values such as 0 that should represent null
> when exposed outside the schema. Below is an example of a scalar intended to represent a database
> autoincrementing numeric identifier internally, where null values are stored as 0.

Author: ```suggestion
Custom scalars process and handle null values during serialization and deserialization. For instance,
let's say you have a database where database ID values of 0 should be represented as graphql 'null'
values.  You could of course write code within the field resolver to translate the values, or you may
utilize a custom scalar to translate the values for you.  See below:
```

Author: Is this better?  I'm not sure.

Reviewer: better
